spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/DatastoreTemplate.java

```diff
@@ -77,23 +78,36 @@ public DatastoreTemplate(Datastore datastore,
 	}
 
 	@Override
-	public <T> void save(T instance) {
-		this.datastore.put(convertToEntity(instance));
+	public <T> T save(T instance) {
+		Entity saved = this.datastore.put(convertToEntity(instance));
+		DatastorePersistentEntity datastorePersistentEntity = this.datastoreMappingContext
+				.getPersistentEntity(instance.getClass());
+		datastorePersistentEntity.getPropertyAccessor(instance).setProperty(
+				datastorePersistentEntity.getIdProperty(), saved.getKey().getNameOrId());
+		return instance;
 	}
 
 	@Override
 	public <T> void deleteById(Object id, Class<T> entityClass) {
 		this.datastore.delete(getKeyFromId(id, entityClass));
 	}
 
+	@Override
+	public <T> void deleteAllById(Iterable<?> ids, Class<T> entityClass) {
+		List<Key> keys = getKeysFromIds(ids, entityClass);
+		this.datastore.delete(keys.toArray(new Key[keys.size()]));
+	}
+
 	@Override
 	public <T> void delete(T entity) {
 		this.datastore.delete(getKey(entity));
 	}
 
 	@Override
-	public void deleteAll(Class<?> entityClass) {
-		this.datastore.delete(findAllKeys(entityClass));
+	public long deleteAll(Class<?> entityClass) {
+		Key[] keysToDelete = findAllKeys(entityClass);
```

Comment:
Since the underlying client library does not return the count, I think we shouldn't do the wasteful find operation, just to get the count. Remember, every operation costs money.

Reply:
This is necessary to get the actual keys to delete. It's NOT just to get the count.